nest/src/token/pUSD.sol

```diff
@@ -2,136 +2,319 @@
-    event Paused(address account);
-    event Unpaused(address account);
 
-    // ========== MODIFIERS ==========
-    modifier whenNotPaused() {
-        require(!paused, "PUSD: paused");
-        _;
-    }
+    // ========== ROLES ==========
+    bytes32 public constant VAULT_ADMIN_ROLE = keccak256("VAULT_ADMIN_ROLE");
 
-    // ========== CONSTRUCTOR & INITIALIZER ==========
-    /// @custom:oz-upgrades-unsafe-allow constructor
+    /**
+     * @notice Prevent the implementation contract from being initialized or reinitialized
+     * @custom:oz-upgrades-unsafe-allow constructor
+     */
     constructor() {
         _disableInitializers();
     }
 
-    function initialize(address _vault, address admin) external initializer {
-        __ERC20_init("", ""); // Empty strings since we override name() and symbol()
-        __AccessControl_init();
-        __UUPSUpgradeable_init();
+    /**
+     * @notice Initialize pUSD
+     * @param owner Address of the owner of pUSD
+     * @param asset_ Address of the underlying asset
+     * @param vault_ Address of the Boring Vault
+     */
+    //
+    function initialize(address owner, IERC20 asset_, address vault_) public initializer {
+        require(owner != address(0), "Zero address owner");
+        require(address(asset_) != address(0), "Zero address asset");
+        require(vault_ != address(0), "Zero address vault");
 
-        vault = IVault(_vault);
+        // Validate asset interface support
+        try IERC20Metadata(address(asset_)).decimals() returns (uint8) { }
+        catch {
+            revert InvalidAsset();
+        }
 
-        _grantRole(DEFAULT_ADMIN_ROLE, admin);
-        _grantRole(UPGRADER_ROLE, admin);
-        _grantRole(MINTER_ROLE, admin);
-        _grantRole(BURNER_ROLE, admin);
-        _grantRole(VAULT_ADMIN_ROLE, admin);
-        _grantRole(PAUSER_ROLE, admin);
-    }
+        super.initialize(owner, "Plume USD", "pUSD", asset_, false, false);
+        __ReentrancyGuard_init();
 
-    // ========== METADATA OVERRIDES ==========
-    function decimals() public pure override returns (uint8) {
-        return 6;
-    }
-
-    function name() public pure override returns (string memory) {
-        return "Plume USD";
-    }
+        pUSDStorage storage $ = _getpUSDStorage();
+        $.vault = IVault(vault_);
 
-    function symbol() public pure override returns (string memory) {
-        return "pUSD";
+        _grantRole(VAULT_ADMIN_ROLE, owner);
     }
 
     // ========== ADMIN FUNCTIONS ==========
+
+    /**
+     * @notice Set a new vault address for the pUSD token
+     * @param newVault Address of the new vault to be set
+     */
     function setVault(
         address newVault
-    ) external onlyRole(VAULT_ADMIN_ROLE) {
-        address oldVault = address(vault);
-        vault = IVault(newVault);
+    ) external nonReentrant onlyRole(VAULT_ADMIN_ROLE) {
+        if (newVault == address(0)) {
+            revert InvalidVault();
+        }
+
```

Comment:
Could add `supportsInterface` check here for ERC-20 (imo unnecessary)

Reply:
setVault removed alltogether from pUSD